opengovsg/FormSG
# feat: add Twilio credentials to forms endpoints

## Problem

Currently, Twilio credentials are directly added via the AWS interface. Instead, we are looking at allowing users to add their twilio credentials to the forms directly through FormSG. 

Related #2155

## Solution

This PR sets up the endpoints that allows the updating and deleting of Twilio credentials
- PUT `admin/forms/:formId/twilio` for
    - creating Twilio credentials
    - updating Twilio credentials
- DELETE `admin/forms/:formId/twilio` to delete Twilio credentials in Secrets Manager


**Features**:
- Creation of updating Twilio credentials routes, controller and services
- Creation of delete Twilio credentials routes, controller and services
- Addition of `SECRET_ENV` environment variable
- Add `TwilioCredentialsData` helper class for parsing and stringification
- Setting up of local MongoDB Replica set to enable Mongo transactions
- Setting up of Secretsmanager on localstack
- Creation of mongo-seed container to seed MongoDB replica set
- Addition of `updateMsgSrvcName` and `deleteMsgSrvcName` instance methods in Form Model
- Addition of `createTwilioTransaction` and `deleteTwilioTransaction` helper functions in `admin-form.service` to handle transactions
- Addition of Joi validator to validate request body of PUT `admin/forms/:formId/twilio`

## Tests

Test cases for the following have been added to the following files to test update and delete twilio credentials functionalities
- `admin-forms.twilio.routes.spec.ts`
- `admin-form.controller.spec.ts`
- `admin-form.service.spec.ts`
- `form.server.model.spec.ts`

Local MongoDB Replica set and SecretsManager (localstack) have been set up for local testing

## Deploy Notes


- **SECRET_ENV variable has to be added to the the configs of AWS staging and production environments before being deployed**
- `?retryableWrite=false` has to be set in production URI string for MongoDB to enable transactions
- IAM role for FormSG has to have new secret manager read, write, delete permissions

## Manual Tests
- New endpoints to test include: 
    - **PUT** `admin/forms/:formId/twilio`
    - **DELETE** `admin/forms/:formId/twilio`
- call **PUT** `admin/forms/:formId/twilio` with invalid twilio credentials
    - Expected: Should throw an error
- call **PUT** `admin/forms/:formId/twilio` with valid twilio credentials
    - Expected: Should be successful
- call **DELETE** `admin/forms/:formId/twilio` on form with stored twilio credentials
      - Expected: Should throw an error
- call **DELETE** `admin/forms/:formId/twilio` on form with no stored twilio credentials
    - Expected: Should be successful (idempotent)
- call **DELETE** `admin/forms/:formId/twilio` on form with stored twilio credentials, follow by a **PUT** `admin/forms/:formId/twilio` to the same form with valid twilio credentials.
    - Expected: Should be successful
- call **PUT** `admin/forms/:formId/twilio` with valid twilio credentials to add twilio credentials. Enter the AWS console and MongoDB (using a GUI) and change the secret name. call **DELETE** `admin/forms/:formId/twilio` on form with stored twilio credentials.
    - Expected: msgSrvcName should be removed from MongoDB while secret in AWS remains.
